vllm/model_executor/models/molmo.py

```diff
@@ -0,0 +1,1130 @@
+                                               QKVParallelLinear,
+                                               RowParallelLinear)
+from vllm.model_executor.layers.layernorm import RMSNorm
+from vllm.inputs import INPUT_REGISTRY, InputContext, LLMInputs
+from vllm.model_executor.models.interfaces import SupportsMultiModal
+from vllm.multimodal import (
+    MULTIMODAL_REGISTRY,
+    MultiModalInputs,
+)
+from vllm.model_executor.layers.sampler import Sampler, SamplerOutput
+from vllm.model_executor.layers.logits_processor import LogitsProcessor
+from vllm.model_executor.layers.quantization.base_config import (
+    QuantizationConfig)
+from vllm.model_executor.layers.rotary_embedding import get_rope
+from vllm.model_executor.layers.vocab_parallel_embedding import (
+    ParallelLMHead, VocabParallelEmbedding)
+from vllm.model_executor.models.utils import make_layers
+from vllm.model_executor.model_loader.weight_utils import default_weight_loader
+from vllm.sequence import (
+    VLLM_TOKEN_ID_ARRAY_TYPE,
+    SequenceData,
+)
+from vllm.model_executor import SamplingMetadata
+from vllm.sequence import IntermediateTensors
+from vllm.transformers_utils.processor import get_processor
+from vllm.platforms import current_platform
+
+log = logging.getLogger(__name__)
+
+
+@dataclass
+class VisionBackboneConfig:
+    image_default_input_size: Tuple[int, int] = (336, 336)
+    image_patch_size: int = 14
+    image_pos_patch_size: int = 14
+    image_emb_dim: int = 1024
+    image_num_heads: int = 16
+    image_num_key_value_heads: int = 16
+    image_num_layers: int = 23
+    image_mlp_dim: int = 4096
+    image_mlp_activations: str = "quick_gelu"
+    image_num_pos: int = 577
+    image_norm_eps: float = 1e-5
+
+    def __post_init__(self):
+        self.image_default_input_size = tuple(self.image_default_input_size)  # type: ignore[assignment]
+
+    @property
+    def image_num_patch(self):
+        h, w = self.image_default_input_size
+        return h // self.image_patch_size, w // self.image_patch_size
+
+
+class ViTMLP(nn.Module):
+    """MLP used in Vision Transformer."""
+    def __init__(
+        self,
+        config: VisionBackboneConfig,
+        quant_config: Optional[QuantizationConfig] = None,
+    ):
+        super().__init__()
+        self.w1 = ColumnParallelLinear(
+            config.image_emb_dim,
+            config.image_mlp_dim,
+            bias=True,
+            quant_config=quant_config,
+        )
+        # Activation function.
+        assert config.image_mlp_activations == "quick_gelu"
+        self.act = QuickGELU()
+        self.w2 = RowParallelLinear(
+            config.image_mlp_dim,
+            config.image_emb_dim,
+            bias=True,
+            quant_config=quant_config,
+        )
+
+    def forward(self, x: torch.Tensor) -> torch.Tensor:
+        x, _ = self.w1(x)
+        x = self.act(x)
+        x, _ = self.w2(x)
+        return x
+
+
+class MultiHeadDotProductAttention(nn.Module):
```

Comment:
I believe it is standard MHA so we should be able to reuse that class.

Reply:
It seems that `CLIPParallelAttention` does not use flash attention.

Reply:
> It seems that CLIPParallelAttention does not use flash attention.

That is correct - currently we use only use xformer/sdpa for the attention module inside the ViT. Empirically switching to flash attention doesn't bring significant benefit to the entire inference pipeline yet since the vision encoder is only run at prefill stage, but using flash attention is indeed on our roadmap!

Reply:
I have cleaned up the code: https://github.com/vllm-project/vllm/pull/9016/commits/72c365e6e3235dbf020a7cfb2643a6506d3cd649, but not using `CLIPParallelAttention` yet. Our team is currently working on an updated version of HF code integrating all Molmo variants, and I will update the vllm code once again after this HF code update.

Reply:
That's fine by me!